### PR TITLE
Allow goals to receive implementations and listeners

### PR DESCRIPTION
**Pull request by cdupuis**
This PR brings infrastructure to create _Typed Goals_: goals that carry their own `ExecuteGoal` implementation. See [Autofix](https://github.com/atomist/sdm/blob/richer-goals/src/api/goal/common/Autofix.ts) for an example. 

The goal will find the SDM instances and register the goal implementations with it, so that from a user API all is left is to do:

```typescript
whenPushSatisfies(AnyPush).setGoals(new Autofix("my-autofix").with(AddThirdPartyLicense));
```

This will allow us to provide a rich ecosystem of Goals for different types of delivery concerns without the requirement to change the core `SoftwareDeliveryMachine` API.

The main changes in this PR are backwards compatible and only additive. Only exception is the move of the well-known-goals to sdm.

### CODE REVIEW

File: src/api-helper/goal/DefaultGoalImplementationMapper.ts [hunk 1]
```diff
@@ -16,51 +16,47 @@
...
-    constructor(private readonly goalLauncher: IsolatedGoalLauncher) {
-    }
-
-    public async findImplementationBySdmGoal(goal: SdmGoalEvent, inv: PushListenerInvocation): Promise<GoalImplementation> {
+    public findImplementationBySdmGoal(goal: SdmGoalEvent): GoalImplementation {
         const matchedNames = this.implementations.filter(m =>
             m.implementationName === goal.fulfillment.name &&
             m.goal.context === goal.externalKey);
 
-        const matchedGoalImplementations = [];
```
Comment (jessitron) [inline src/api-helper/goal/DefaultGoalImplementationMapper.ts]: does this mean there are no more push tests on implementations? ... or that they aren't used?

File: src/api-helper/goal/storeGoals.ts [hunk 2]
```diff
@@ -115,8 +116,8 @@ export function constructSdmGoal(ctx: HandlerContext, parameters: {
     url?: string,
     fulfillment?: SdmGoalFulfillment,
 }): SdmGoalMessage {
-    const {goalSet, goal, goalSetId, state, id, providerId, url} = parameters;
-    const fulfillment = parameters.fulfillment || {method: "other", name: "unspecified"};
+    const { goalSet, goal, goalSetId, state, id, providerId, url } = parameters;
+    const fulfillment = parameters.fulfillment || { method: SdmGoalFulfillmentMethod.Other, name: "unkown" };
```
Comment (jessitron) [inline src/api-helper/goal/storeGoals.ts]: spelling: unkown

File: src/api-helper/goal/DefaultGoalImplementationMapper.ts (hunk 1 above)
Comment (cdupuis) [reply src/api-helper/goal/DefaultGoalImplementationMapper.ts]: no, they are being used and evaluated when we SetOnGoalsOnPush. I added this here again last week which was a mistake.

File: src/api-helper/goal/storeGoals.ts (hunk 2 above)
Comment (cdupuis) [reply src/api-helper/goal/storeGoals.ts]: Good point.

File: src/api-helper/machine/AbstractSoftwareDeliveryMachine.ts [hunk 3]
```diff
@@ -268,6 +275,8 @@ export abstract class AbstractSoftwareDeliveryMachine<O extends SoftwareDelivery
                           goalSetters: Array<GoalSetter | GoalSetter[]>) {
         super();
         // If we didn't get any goal setters don't register a mapping
+        registrableManager().register(this);
```
Comment (jessitron) [inline src/api-helper/machine/AbstractSoftwareDeliveryMachine.ts]: what does the global __registrable do?

Comment (johnsonr) [reply src/api-helper/machine/AbstractSoftwareDeliveryMachine.ts]: it should be `registerable`

**Pull request merged**